File: app/app/api/routes/location_sync.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.country import Country
from app.models.state import State
from app.models.city import City

router = APIRouter(
    prefix="/location-sync",
    tags=["Location Sync"],
    # Require login for every endpoint in this router
    dependencies=[Depends(get_current_user)],
)


# Test endpoint
@router.get("/test")
def test():
    return {"message": "location sync working"}


# Add country manually
@router.post("/countries")
def create_country(name: str, db: Session = Depends(get_db)):

    existing = db.query(Country).filter(Country.name == name).first()

    if existing:
        return {"message": "Country already exists"}

    country = Country(name=name)

    db.add(country)
    db.commit()
    db.refresh(country)

    return {
        "message": "Country created",
        "id": country.id,
        "name": country.name
    }


# Get all countries
@router.get("/countries")
def get_countries(db: Session = Depends(get_db)):

    countries = db.query(Country).all()

    return countries


# Add state
@router.post("/states")
def create_state(name: str, country_id: int, db: Session = Depends(get_db)):

    state = State(
        name=name,
        country_id=country_id
    )

    db.add(state)
    db.commit()
    db.refresh(state)

    return state


# Add city
@router.post("/cities")
def create_city(name: str, state_id: int, db: Session = Depends(get_db)):

    city = City(
        name=name,
        state_id=state_id
    )

    db.add(city)
    db.commit()
    db.refresh(city)

    return city


# Get all cities
@router.get("/cities")
def get_cities(db: Session = Depends(get_db)):

    return db.query(City).all()

File: frontend/src/api/locationsync.js
// src/api/locationSyncAPI.js

import api from "./axios";

/* =====================================================
   COMMON ERROR HANDLER
===================================================== */
const handleRequest = async (request) => {
  try {
    const res = await request;
    return res.data;
  } catch (error) {
    console.error(
      "Location Sync API Error:",
      error?.response?.data || error.message
    );

    throw new Error(
      error?.response?.data?.detail ||
      error?.response?.data ||
      "Location sync request failed"
    );
  }
};

/* =====================================================
   HEALTH CHECK (OPTIONAL)
   GET /location-sync/test
===================================================== */
export const locationSyncTest = () =>
  handleRequest(api.get("/location-sync/test"));

/* =====================================================
   🌍 COUNTRIES
===================================================== */

/* GET ALL COUNTRIES */
export const getCountries = () =>
  handleRequest(api.get("/location-sync/countries"));

/* CREATE / SYNC COUNTRY */
export const createCountry = (name) =>
  handleRequest(
    api.post("/location-sync/countries", {
      name,
    })
  );

/* =====================================================
   🗺️ STATES
===================================================== */

/* CREATE STATE */
export const createState = (name, countryId) =>
  handleRequest(
    api.post("/location-sync/states", {
      name,
      country_id: countryId,
    })
  );

/* =====================================================
   🏙️ CITIES
===================================================== */

/* GET ALL CITIES */
export const getCities = () =>
  handleRequest(api.get("/location-sync/cities"));

/* CREATE CITY */
export const createCity = (name, stateId) =>
  handleRequest(
    api.post("/location-sync/cities", {
      name,
      state_id: stateId,
    })
  );